### PR TITLE
fix: tree item sort order in treeview v1 to be on par with v2 in preparation for v2 deprecation

### PR DESCRIPTION
**Pull request by hikchoi**
# fix: fix item sorting in treeview v1 to be on par with v2

This PR:
- Fixes an issue with tree view v1 not sorting notes properly.
  - children were being sorted, but not being set (`_.sortBy` is does not mutate the object.)
  - when the tree view is first caculated, the sort order is _sorta_ right (ordered by title, presumably by the engine elsewhere when the children were added / retrieved / indexed), but a rename operation would have broke the ordering because of this issue.
- Fixes an issue with `TreeUtils.sortNotesAtLevel` when bubbling down `tags` hierarchy.
  - If a note with title that comes after `T` exists, the `tags` hierarchy will not be bubbled down.
  - This fixes the alphabetical ordering, and it now respects the `nav_order` property if it exists in `tags.md`.
- Uses `TreeUtils.sortNotesAtLevel` for sorting items
  - will respect `nav_order`
  - will then sort by title, then by updated timestamp
  - and push down tags to the bottom of the tree if `nav_order` is not set.
- Explicitly sets the label of the tree view item.
  - before, this was not set unless it was a root note and in this case the label is derived from the resource URI
  - This caused some confusion since the ASCII ordering uses the title which is capitalized most of the time, but the label shows the uncapitalized basename of the file.
    -  (i.e. notes with title `A` and file name `a.md` came before a note with title `_b` and file name `_b.md` because underscore is _after_ uppercase alphabets in the ASCII chart.)
- ~~Adds a test wrapper `testAsyncWithCallback`, that allows to overspecify the test resolution.~~
  - ~~Can be used with `describeMulti/SingleWS` harness when you need to run an async test and also explicitly call `done`~~
  - ~~daily async created for this~~
- Refactor one existing test in `EngineNoteProvider.test.ts`
  - to use the new test harness
  - ~~to use the `testAsyncWithCallback` wrapper instead~~
  - ~~and prevent `done` from being called twice when failed~~

### Basics

- [x] code should follow [Code Conventions](https://wiki.dendron.so/notes/773e0b5a-510f-4c21-acf4-2d1ab3ed741e.html)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](https://wiki.dendron.so/notes/pMS27sHxbWeKMoPRrWEzs.html).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics

- [~] if you are adding analytics related changes, make sure you [Document](https://wiki.dendron.so/notes/8ThPaB9iXXm2Szk3C9kFt.html) changes in airtable

### Extended

- [~] can we track the performance of this change to know if it is _successful_? 
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## 

### Basics

- [x] [Write Tests](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] [Confirm existing tests pass](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)
- [x] [Confirm manual testing](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)

### Extended

- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](https://wiki.dendron.so/notes/dtMsF12SF2SUhLN10sYe2.html)

- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome

### Basics

- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 

### Basics

- [x] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](https://wiki.dendron.so/notes/3489b652-cd0e-4ac8-a734-08094dc043eb.html) or [Packages](https://wiki.dendron.so/notes/32cdd4aa-d9f6-4582-8d0c-07f64a00299b.html)

## 

### Extended

- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](https://wiki.dendron.so/notes/iZ8vpfY0n1E3Nq3IC1Y7X.html)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

**Comment by hikchoi** [comment]
@jonathanyeung Trying to add tests for this in [[../packages/plugin-core/src/test/suite-integ/EngineNoteProvider.test.ts]] but I'm a bit stuck here.

The existing tests are using the legacy harness, so we can pass in `done` and also have `onInit` async and call `done` with an error.

With our new harness (`describeMultiWS`), we can't overspecify the resolution. We either make it async and pass `done` or keep it sync and pass `done`. How would you change the tests here if we were to refactor them to use the new test harness?

**Comment by jonathanyeung** [comment]
@hikchoi - I think something like this should work? 

```typescript
suite("EngineNoteProvider Tests", function testSuite() {
  // Set test timeout to 2 seconds
  this.timeout(2000);

  describeMultiWS(
    "WHEN workspace type is not specified",
    {
      preSetupHook: ENGINE_HOOKS.setupBasic,
    },
    () => {
      test("THEN initializes correctly", async () => {
        const { wsRoot, vaults } = getDWorkspace();

        const testNoteProps = await NoteTestUtilsV4.createNote({
          fname: "alpha",
          vault: vaults[0],
          wsRoot,
        });

        const mockEvents = new MockEngineEvents();
        const provider = new EngineNoteProvider(mockEvents);

        const promise = new Promise<void>((resolve) => {
          provider.onDidChangeTreeData(() => {
            // If you comment line below, test should fail.
            resolve();
          });
        });

        const entry: NoteChangeEntry = {
          note: testNoteProps,
          status: "create",
        };

        mockEvents.testFireOnNoteChanged([entry]);

        return promise;
      });
    }
  );
  ```

**Comment by SeriousBug** [comment]
I think you could also use the new harnesses with a test function that uses the `done` callback.

**Comment by hikchoi** [comment]
> @hikchoi I think you could also use the new harnesses with a test function that uses the `done` callback.

yes, but we want the tests to be async here, and we can't use the done callback if it's async. Am I understanding that wrong? 😕

**Comment by SeriousBug** [comment]
@hikchoi You could use a trick like this:
```ts
test("THEN something", (done) => {
  (async () => {
    // ...
    done();
  })();
});
```
It's a bit ugly, but functionally equivalent to how the old test harnesses worked. We can add a helper function like `testAsyncWithCallback` to make this a little nicer.

File: packages/plugin-core/src/test/suite-integ/EngineNoteProvider.test.ts [hunk 1]
```diff
@@ -1,169 +1,317 @@
...
+                  done(error);
                 }
-                break;
-              }
-              default: {
-                done({ message: "Note with unexpected vault found!" } as Error);
               }
+            });
+
+            if (!error) done();
```
Comment (hikchoi) [inline packages/plugin-core/src/test/suite-integ/EngineNoteProvider.test.ts:191]: `done` was being called twice when it failed which made the error message hard to read.

**Review by jonathanyeung — APPROVED** [review]
lgtm other than a merge conflict with `yarn.lock`

**Review by kevinslin — APPROVED** [review]

**Pull request merged**